megamek/src/megamek/common/pathfinder/BoardClusterTracker.java

```diff
@@ -0,0 +1,310 @@
+                                entity.hasWorkingMisc(MiscType.F_FULLY_AMPHIBIOUS) ||
+                                entity.hasWorkingMisc(MiscType.F_LIMITED_AMPHIBIOUS);
+        
+        for(int x = 0; x < board.getWidth(); x++) {
+            for(int y = 0; y < board.getHeight(); y++) {
+                Coords c = new Coords(x, y);
+                
+                // hex is either inaccessible
+                // or it is inaccessible AND we can't level it, then we move on
+                if (entity.isLocationProhibited(c) && 
+                        (!destructionAware || (destructionAware && !canLevel(entity, c)))) {
+                    continue;
+                }
+                
+                List<Coords> neighborsToJoin = new ArrayList<>();
+                BoardCluster biggestNeighbor = null; 
+                
+                // hex is accessible one way or another
+                for(int direction = 0; direction < 6; direction++) {
+                    Coords neighbor = c.translated(direction);
+                    
+                    if(clusters.containsKey(neighbor)) {
+                        int neighborElevation = BoardEdgePathFinder.calculateUnitElevationInHex(board.getHex(neighbor), entity, isHovercraft, isAmphibious);
+                        int myElevation = BoardEdgePathFinder.calculateUnitElevationInHex(board.getHex(c), entity, isHovercraft, isAmphibious);
+                        
+                        // if we can't reach from here to the neighbor due to elevation differences, move on
+                        int elevationDiff = Math.abs(neighborElevation - myElevation);
+                        if(elevationDiff > entity.getMaxElevationChange()) {
+                            continue;
+                        }
+                        
+                        // we can "freely" go back and forth between ourselves and the neighbor, so let's join that cluster 
+                        neighborsToJoin.add(neighbor);
+                        
+                        if((biggestNeighbor == null) || 
+                                (clusters.get(neighbor).contents.size() > biggestNeighbor.contents.size())) {
+                            biggestNeighbor = clusters.get(neighbor);
+                        }
+                    }
+                }
+                
+                // start up a new cluster if we have no mutually accessible neighbors
+                if(neighborsToJoin.isEmpty()) {
+                    BoardCluster newCluster = new BoardCluster(clusterID++);
+                    newCluster.contents.add(c);
+                    clusters.put(c, newCluster);
+                // otherwise, join an existing cluster, bringing any other mutually accessible neighbors and their clusters with me
+                // join the biggest neighbor to reduce shuffling.
+                } else {
+                    biggestNeighbor.contents.add(c);
```

Comment:
This may be null, especially if elevation changes prevent any neighbor to be added (perhaps they jumped into a hole).

Reply:
I think it's actually not possible, because if it's null then neighborsToJoin is empty, and then it goes into the "new cluster" logic, but I'll add it to that clause out of an abundance of caution.